src/coreclr/jit/importer.cpp

```diff
@@ -4396,14 +4399,11 @@ GenTree* Compiler::impImportStaticFieldAccess(CORINFO_RESOLVED_TOKEN* pResolvedT
     }
```

Comment:
`FieldSeqStore::Create` seems to expect that given field is always accessed the same way (it has `assert(fieldSeq->GetOffset() == offset)`).

A static constructor may finish running between first and second access of a field in given method. I expect that it will cause the assert in `FieldSeqStore::Create` to fire, and potentially more things can go wrong after that.

Reply:
Right

Reply:
Removed that check and checked that JIT carries "requires static init" flag if needed.

Reply:
Is it really fixing the race condition? The initialized vs. non-initialized check is effectively done by the `frozenObj != nullptr` condition. I think the race condition is still there.

Reply:
I see, can we pre-allocate statics in getFieldInfo so we can remove the !=nullptr check?

Reply:
Possibly, but it won't be a simple one-line change.

In general, it is not a good idea to make assumptions in the JIT that subsequent getCallInfo or getFieldInfo calls are going to return same pointers for given method or field.

Reply:
`FieldSeqStore` is a dictionary like that. Can we reuse it for this purpose?

Reply:
Ah, so jit sees that something has changed, and re-sends `getFieldInfo` request with "don't give me frozen object" flag?

Reply:
> so jit sees that something has changed, and re-sends getFieldInfo request with "don't give me frozen object" flag?

I do not think that the JIT should be repeating any requests hoping for better answer. If it gets two different answers to the same question at two points in time, it can pick the answer that it likes better.

Reply:
> pick the answer that it likes better

It is not possible under the current scheme (well... not easily at least, one could imagine reimporting earlier blocks). Field sequences are tied to the IR shape produced for a given field (they "describe" it).

It is possible to tie sequences to specific IR locations instead, but it would not be a simple change, and it would come with throughput regressions.

Reply:
Then just a separate global (per-jit) `hashmap<FIELD_HANDLE, void*>` structure that Jit will consult with?

Reply:
It sounds like fixing the JIT/EE interface to always return the same answer in this case may be actually simpler...

Reply:
Added a cache. Tests seem to be still passing. Jit-diff is the same

Reply:
I thought that you plan to add this cache on the JIT side. This cache does not belong to the EE side. If you want to fix it on the EE side, we should do the larger fix to initialize boxes as needed, so that the JIT/EE interface always gives the same answer.

Reply:
I can move it to jit, it's just I was confused by your "fixing the JIT/EE interface to always return the same answer in this case may be actually simpler" - I though you meant that file.

Reply:
I meant creating the boxes on demand on the EE side. Some changes for that would be in jitinterface.cpp and some changes would be in other files. We would need to synchronize with code that creates the boxes created right before we run the static constructor (we do not want the boxes created by one to overwrite the boxes created by the other).

Reply:
I would do the proper fix with allocating the boxes on demand. It should not be too bad as I am thinking about it.

Reply:
Alright, did the lazy initialization (guarded with a lock)